ambry-account/src/main/java/com/github/ambry/account/AccountMetadataStore.java

```diff
@@ -0,0 +1,107 @@
+ */
+abstract class AccountMetadataStore {
+  private static final Logger logger = LoggerFactory.getLogger(AccountMetadataStore.class);
+
+  protected final AccountServiceMetrics accountServiceMetrics;
+  protected final LocalBackup backup;
+  protected final String znRecordPath;
+  private final HelixPropertyStore<ZNRecord> helixStore;
+
+  /** Create a new {@link AccountMetadataStore} instance for the subclasses.
+   * @param accountServiceMetrics The {@link AccountServiceMetrics}
+   * @param backup The {@link LocalBackup} to manage the backup files.
+   * @param helixStore The {@link HelixPropertyStore} to retrieve and update the {@link ZNRecord}.
+   * @param znRecordPath The {@link ZNRecord} path.
+   */
+  AccountMetadataStore(AccountServiceMetrics accountServiceMetrics, LocalBackup backup,
+      HelixPropertyStore<ZNRecord> helixStore, String znRecordPath) {
+    this.accountServiceMetrics = accountServiceMetrics;
+    this.backup = backup;
+    this.helixStore = helixStore;
+    this.znRecordPath = znRecordPath;
+  }
+
+  /**
+   * ZKUpdater extends the {@link DataUpdater} with another method to  permform some clean up logic after
+   * an update.
+   */
+  interface ZKUpdater extends DataUpdater<ZNRecord> {
+
+    /**
+     *  Called after {@link HelixPropertyStore} update the {@code znRecordPath} with the {@link ZKUpdater}.
+     * @param isUpdateSucceeded The result of the {@link HelixPropertyStore#update}.
+     */
+    void afterUpdate(boolean isUpdateSucceeded);
+  }
+
+  /**
+   * Fetch the {@link Account} metadata from the given ZNRecord. It should return null when there is no {@link Account}
+   * ever created before. Subclass can assume the {@link ZNRecord} passed in this function is not null;
+   * @param record The {@link ZNRecord} fetched from {@code znRecordPath}.
+   * @return {@link Account} metadata in a map.
+   */
+  abstract Map<String, String> fetchAccountMetadataFromZNRecord(ZNRecord record);
+
+  /**
+   * Create new {@link ZKUpdater} that will be used to update the accounts.
+   * @param accounts The {@link Account} collection to update.
+   * @return the Specific {@link ZKUpdater} to update accounts with {@link HelixPropertyStore}.
+   */
+  abstract ZKUpdater createNewZKUpdater(Collection<Account> accounts);
+
+  /**
+   * fetchAccountMetadata would fetch the latest full set of {@link Account} metadata from the store. It returns null
+   * when there is no {@link Account} created.
+   * @return {@link Account} metadata in a map.
+   */
+  Map<String, String> fetchAccountMetadata() {
+    long startTimeMs = System.currentTimeMillis();
+    logger.trace("Start reading ZNRecord from path={}", znRecordPath);
+    ZNRecord znRecord = helixStore.get(znRecordPath, null, AccessOption.PERSISTENT);
+    logger.trace("Fetched ZNRecord from path={}, took time={} ms", znRecordPath, startTimeMs);
+    if (znRecord == null) {
+      logger.debug("The ZNRecord to read does not exist on path={}", znRecordPath);
```

Comment:
but it's not a warning message. ZNRecord can be null here. It's one of the legit state.

Reply:
If not `warn`, I would suggest `info` because such info is supposed to be printed out (`debug` level is turned off by default).  I feel like ZNRecord should exist in almost all cases, if not, it's actually a warning that either the ZNode is deleted or znRecordPath is incorrect.